test/pylib/scylla_cluster.py

```diff
@@ -1301,12 +1306,19 @@ async def _after_test(self, _request) -> str:
         # copy current tasks
         tasks = [key for key in self.tasks_history.keys()]
         # wait for all other tasks in ScyllaClusterManager
-
-        for task in tasks:
-            request = self.tasks_history.pop(task)
-            if not task.done():
-                self.logger.info("wait for task:%s, request:%s", task, request.path_qs)
-                await asyncio.wait_for(task, timeout=120)
+        try:
+            for task in tasks:
+                request = self.tasks_history.pop(task)
+                if not task.done():
+                    self.logger.info("wait for task:%s, request:%s", task, request.path_qs)
+                    await asyncio.wait_for(task, timeout=120)
+        except asyncio.TimeoutError:
+            self.break_manager(f"error on waiting coro {task.get_name()}")
+
+        # check on tasks leakage
+        await asyncio.sleep(0.1)
+        if self.tasks_history:
+            self.break_manager(f"tasks leakage found  {self.tasks_history}")
```

Comment:
This solution doesn't look reliable... You can always have a runaway asyncio task which waits for a long time and then sends some requests. I don't think this will catch it.

I think it would be better if every test case received its own instance of the ManagerClient (as far as I know right now we are reusing it for multiple test cases). At the end of the test, we could gracefully stop it, preventing more requests from being sent and waiting for the existing ones to complete.

If sharing some resources (e.g. the HTTP client) is important, then the instances of the manager client should be able to do it.

I know that, technically speaking, a test case could create its own manager instance by itself and this mechanism wouldn't catch this. However, I believe that such mistake is much less likely to happen and is much more likely to be spotted.

@kostja , @kbr-scylla - do you perhaps know if there is a good reason why we reuse the same ManagerClient instance for multiple tests?

Reply:
> @kostja , @kbr-scylla - do you perhaps know if there is a good reason why we reuse the same ManagerClient instance for multiple tests?

ManagerClient corresponds to a ScyllaCluster (1-1 correspondence) and clusters can in theory be reused between test cases (if they are not left dirty and are in the same suite)

Reply:
> ManagerClient corresponds to a ScyllaCluster (1-1 correspondence) and clusters can in theory be reused between test cases (if they are not left dirty and are in the same suite)

Hm, but that doesn't explain it. I'm a bit confused as well.

The scope of `manager` fixture is `function`:
```python3
@pytest.fixture(scope="function")
async def manager(request, manager_internal):
```

internally it uses `manager_internal` with the scope `session`
```python3
@pytest.fixture(scope="session")
async def manager_internal(event_loop, request):
```

IIUC we have different `pytest` executions for different test files, i.e., different test files == different sessions.

But clusters can be reused between different test files (?), since they are all accessing the same pool living in test.py process.

So different `ManagerClient`s can be used to reuse a cluster. (Need to confirm with experiment that it is indeed the case.)

So we should be able to change scope of `manager_internal` to be `function` and then we would have separate `ManagerClient`s.

But that is only the client-side, i.e. the `pytest` process. While the problematic part is the server-side, `ScyllaClusterManager` living inside the `test.py` process. The code we're looking at here is the server-side. Could we move all of this logic to the client-side instead (`ManagerClient`)?

If not, we'd also have to change `ScyllaClusterManager`, which IIUC currently is also per test file, not per test case.

That would multiply the amount of resources (like sockets), but also perhaps allow us to run test cases in the same test file concurrently. (Which is a long-requested feature.)

Reply:
> If not, we'd also have to change ScyllaClusterManager, which IIUC currently is also per test file, not per test case.

This'd have to be done anyway if we made `ManagerClient` function-scope, because two `ManagerClient`s cannot use a single `ScyllaClusterManager` concurrently.

Reply:
@kbr-scylla @piodul 
client side fix for case of 
``` having a runaway asyncio task which waits for a long time and then sends some requests.```
https://github.com/scylladb/scylladb/pull/18563

Reply:
added 2 more commits to fixe described problem